packages/ckeditor5-autoformat/src/blockautoformatediting.js

```diff
@@ -82,18 +85,34 @@ export default function blockAutoformatEditing( editor, plugin, pattern, callbac
 
 		const blockToFormat = entry.position.parent;
 
-		// Block formatting should trigger only if the entire content of a paragraph is a single text node... (see ckeditor5#5671).
-		if ( !blockToFormat.is( 'paragraph' ) || blockToFormat.childCount !== 1 ) {
+		// Block formatting should be disabled in codeBlocks (#5800).
+		if ( blockToFormat.is( 'codeBlock' ) ) {
 			return;
 		}
 
-		const match = pattern.exec( blockToFormat.getChild( 0 ).data );
+		// In case a command is bound, do not re-execute it over an existing block style which would result with a style removal.
+		// Instead just drop processing so that autoformat trigger text is not lost. E.g. writing "# " in a level 1 heading.
+		if ( command && command.value === true ) {
+			return;
+		}
+
+		const firstNode = blockToFormat.getChild( 0 );
+		const match = pattern.exec( firstNode.data );
 
 		// ...and this text node's data match the pattern.
 		if ( !match ) {
 			return;
 		}
 
+		const range = first( editor.model.document.selection.getRanges() );
```

Comment:
I don't understand why you're changing this. Why did you remove `$` from the regexps?

Reply:
Without tailing `$` removed it would only match blocks with text noes that contain _only_ the autocomplete trigger character. While related issue requests that non empty blocks like `###^ foo` should also be transformed.

And for the line referenced by this comment, the range check is to only trigger autocompletion when the change occurred in `### ^foo` and not in another place, like `### fo^o` - you don't want to trigger autocomplete when typing in the middle of the block (also handles #5671 referenced in the comment below).

Reply:
What if the string that we match the regexp to is trimmed to the current selection position. Does it make sense for it to contain characters on the right side of the selection?

Reply:
We're scanning the text from the very first text node anyway. If there's an element after that text node, or another text node (different attributes), then we'd not try to match it against the given regexps anyway.

Therefore, we can't reliably say that you can also match to the right of the caret. We're looking for the text on the left of the caret. Therefore, in this case:

```
<p>## ^xx</p>
```

In my opinion, we should be matching "## " only, so the regexps could look like this: `/^## $/`.

Right now you're matching "## xx" against `/^## /` and then checking if that occurred right before the selection which is counterintuitive to me. What's also counterintuitive is that you won't be matching against the entire element content anyway – only the first text node is used as I mentioned. That's not very straightforward.

However, it's not a big deal either so if you want to keep this implementation no big harm will be done. I'll just leave comments how you can improve it.

Reply:
`selection.getFirstPosition()` instead of what you're doing this + check before we do ANYTHING that the selection is collapsed.

Reply:
Restored original regexp format in a63f559.